### PR TITLE
Implement auto-downloading of IERS-A data

### LINKED ISSUE #3275: Automatic download of IERS A table for times not covered by IERS B
It has turned out to be very confusing to users (and us!) that `Time` cannot calculate the sidereal time for times around now or in the near future. For this purpose, the IERS A table can be used with some handy-work [1], but this could and probably should be automated. One question is the extent to which this should be completely automatic -- the file is not exactly small.  Another question is what should happen for times further in the future. Possibly `sidereal_time` at least could have an option `sloppy=True` where future leap seconds and earth rotation are ignored. 

[1] http://docs.astropy.org/en/latest/time/index.html#transformation-offsets

### PR DESCRIPTION
**Pull request by taldcroft**
This is a candidate implementation to resolve #3275.

I'm putting this up prior to writing tests or even having it pass existing tests because it may be controversial.  :smile:  I've been thinking about this over the last week and have started at least four different versions.  The trick is maintaining possibly strict back-compatible behavior (for the minority with this requirement) while introducing a default, *SIMPLE* behavior that just works for the majority.  I find the existing IERS a bit confusing because of all the options that are supported.

The key is defaulting to auto-downloading and using IERS-A when required, by use of a new class `IERS_Auto`.  The exact caching mechanism is TBD, but as mentioned we *know* in advance exactly when the source file is updated, so it is a really easy problem in this case.  In this implementation I just put in a config var that specifies how old the cached file can be.  Don't worry about this bit.  However, I will say that I thought about fine-tuned optimizations like @mhvk suggested in #3275 and in the end rejected them as just too complicated.  I think it's better to have a simple scheme that can be explained and understood by users and developers in a sentence or two.

What is more important is automatically using IERS-A via the `IERS_Auto` class in `Time` and `Coordinates`.  This means that when a user first requires IERS values the file will always get downloaded.  Supplying an IERS-A file with astropy is a bit pointless because it will always get superceded very shortly.  

The auto-download behavior can be fully reverted to the current 1.1 behavior by setting a config var `astropy.utils.data.iers.conf.auto_download` to False (default = True).  Users that require the legacy strict IERS-B can do this.

Another important idea: the `IERS_Auto` class is exempted from range checking because it is assured of having the most recent available data.  This means the (possibly-clipped) interpolation is always within the confidence bounds of current Earth rotation knowledge.  For actual use cases in astronomy (particularly observation planning) everything just works as well as it can given our best knowledge of Earth rotation.  All the adhoc workarounds currently suggested would go away.

Of course the downside is that (in the default behavoir) reproducibility to numerical precision would go away.  But someone would be free to write a script with auto-download disabled and thus get reproducible behavior.  And in all cases reproducibility to within confidence limits of Earth rotation parameters would not be affected.

### CODE REVIEW

**Comment by taldcroft** [comment]
cc: @embray @eteq

**Comment by taldcroft** [comment]
And of course this needs to smoothly handle the case where the network turns out to be unavailable.  First, use a timeout for the initial download of IERS-A.  If this times out, then fall through to the standard IERS-B with a warning.  (Note that in this case any extrapolation will generate an exception).  Likewise if the currently cached IERS-A is too old and then the subsequent download of IERS-A fails, also fall through to IERS-B with a warning (which would also say how old the current file is).

File: astropy/utils/iers/iers.py [hunk 1]
```diff
@@ -443,5 +473,49 @@ def pm_source(self, i):
         """Set PM source flag for entries in IERS table"""
         return np.ones_like(i) * FROM_IERS_B
 
+class IERS_Auto(IERS_A):
```
Comment (embray) [inline astropy/utils/iers/iers.py]: I don't understand why a separate class instead of a new argument to the existing class's `open` method.

**Comment by embray** [comment]
I still feel like if someone wanted to tackle #3961 that would make this mostly moot, but this is simpler for now...

File: astropy/utils/iers/iers.py (hunk 1 above)
Comment (taldcroft) [inline astropy/utils/iers/iers.py]: I did think about that.  My first issue is that the `IERS` class is already pretty confusing to me, so adding an option that completely changes the behavior (and in fact the *class* of the output) of `open()` is just too much.

The other problem that convinced me a new class is the best way is that all of the existing options to `IERS.open()` are moot in this case.  Thus you would need to raise an exception if any args or kwargs *besides* the `auto_update` option were provided.  That just seems like a bad API.

**Comment by taldcroft** [comment]
> I still feel like if someone wanted to tackle #3961 that would make this mostly moot, but this is simpler for now...

The caching is the easy part of this, and if someone does #3961 then that will be quickly inserted in here.

The hard part (conceptually) is plugging a new default behavior deep into Time and Coordinates without screwing things up, while allowing for back-compatibility.

**Comment by taldcroft** [comment]
One more thing is that it wouldn't be too hard to have "on-demand" downloading which is triggered by actually trying to do a conversion that is in a time range where newer data would help.  The exact algorithm for this would depend on details of the IERS bulletin files.  This tighter coupling is not ideal, but on the whole probably worth doing to prevent unnecessary network access.

File: astropy/utils/iers/iers.py (hunk 1 above)
Comment (embray) [inline astropy/utils/iers/iers.py]: Yeah, I thought that might be your thinking, and that makes sense.

If it were possible I would change the existing hierarchy such that there's some hidden `_IERSBase` class that is the same as the current `IERS` class, and what you're calling `IERS_Auto` would become just `IERS`.  I wonder if it isn't too late to make that change since direct use of `IERS` isn't particularly advertised in the docs.

**Comment by taldcroft** [comment]
@mhvk @eteq - I have done a bunch of work on the auto-download of IERS data and I think this is getting very close.  This PR is based on #4810 because it won't work without that.

I know this is not going to be entirely straightforward to review, but please plan to carve out some time in advance of the code freeze.  Some of the logic behind the behavior is a bit intricate, but I've tried to be fairly verbose in the code comments.

Some known issues:

- [ ] Test with iers.conf.max_auto_age = None.
- [ ] I was getting one doc build warning that I did not understand.
- [ ] The doc section on Auto-refresh behavior is not finished.
- [ ] Time and coordinate docs should be updated to provide cross reference links.

File: docs/utils/iers.rst [hunk 2]
```diff
@@ -0,0 +1,165 @@
+.. _utils:
```
Comment (pllim) [inline docs/utils/iers.rst]: @taldcroft , if you are talking about `.../utils/index.rst:5: WARNING: duplicate label utils, other instance in .../utils/iers.rst` Sphinx warning, it means that this label has been used. Change it to something more unique like `.. _utils-iers:` would fix the warning. Of course, you also have to change any references to this particular label (but not the other conflicting one).

Comment (taldcroft) [inline docs/utils/iers.rst]: @pllim - thanks!  I got sidetracked about the error it was reporting and usually expect Sphinx messages to be misleading or useless.  In this case though I should have been able to see it.

**Comment by taldcroft** [comment]
OK, now something strange has come up.  It looks like the IERS-B values that are supplied in the `finals2000A.all` file (http://maia.usno.navy.mil/ser7/finals2000A.all) do not match corresponding values from the actual IERS-B file (http://hpiers.obspm.fr/iers/eop/eopc04/eopc04_IAU2000.62-now).  For instance in the table below columns 2 and 3 are from IERS-B, while column 4 (UT1_UTC_B_finals) is from the UT1_UTC_B column of the IERS-A file.  The last column `delta` is the difference.  As you can see the `delta` value can be up to 5 times larger than then stated error on UT1-UTC, i.e. 18 microsec vs. 3.4 microsec.
```
In [41]: iers_b_ok
Out[41]: 
<IERS_B length=7>
  MJD    UT1_UTC  e_UT1_UTC UT1_UTC_B_finals       delta       
   d        s         s            s                 s         
float64  float64   float64      float64           float64      
------- --------- --------- ---------------- ------------------
53747.0 0.3335044  1.03e-05         0.333506 -1.60000000005e-06
53748.0 0.3337777   3.4e-06         0.333768            9.7e-06
53749.0 0.3339995   6.3e-06         0.334007 -7.49999999999e-06
53750.0 0.3340951   7.4e-06         0.334111          -1.59e-05
53751.0 0.3340757   5.5e-06         0.334096          -2.03e-05
53752.0 0.3339359   3.4e-06         0.333954          -1.81e-05
53753.0 0.3336386   1.4e-06         0.333636  2.60000000002e-06
```
This came up in regression testing.
```
724   >>> import astropy.units as u
725   >>> t = Time('2006-01-15 21:24:37.5', format='iso', scale='utc',
726   ...          location=(-155.933222*u.deg, 19.48125*u.deg), precision=6)
727   >>> t.utc.iso
728   '2006-01-15 21:24:37.500000'
729   >>> t.ut1.iso
Expected:
    '2006-01-15 21:24:37.834078'
Got:
    '2006-01-15 21:24:37.834098'
```

**Comment by mhvk** [comment]
Yikes, that is not good at all! Though now that you mention it, I vaguely recall that I saw that the two columns were not identical before (it is already strange that they have different number of significant digits, though the errors you find are clearly larger than this). There may be errors in initial versions of IERS B that are corrected but not propagated to IERS A. Sigh.

**Comment by taldcroft** [comment]
So who is right?  It's easy enough to use the actual available IERS-B values (that are bundled with astropy) in IERS_Auto.

**Comment by mhvk** [comment]
IERS B is supposed to have final values.

**Comment by taldcroft** [comment]
@mhvk - see 09c02a6 for a patch.  Ugh.

File: astropy/coordinates/tests/accuracy/test_altaz_icrs.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def test_against_hor2eq():
     # radec_expected = ra=3.30875 deg, dec=15.183416666666666 deg
     # radec_actual = ra=3.3094193224314625 deg, dec=15.183757021354532 deg
     # distance = 2.6285 arcsec
-    assert distance < 5 * u.arcsec
+    assert distance < 10 * u.arcsec
```
Comment (mhvk) [inline astropy/coordinates/tests/accuracy/test_altaz_icrs.py]: Did this get worse? That seems very strange!

**Comment by mhvk** [comment]
@taldcroft - I've started looking at this.  Overall, it looks good, though I want to think a bit more whether it isn't possible to make more minimal changes to `IERS` itself, keeping additions mostly within the `IERS_Auto` class.

One annoyance: now that `IERS_A` turns out not to have correct `IERS_B` values, shouldn't we auto-update `IERS_B` too?!

File: astropy/coordinates/tests/accuracy/test_altaz_icrs.py (hunk 3 above)
Comment (eteq) [inline astropy/coordinates/tests/accuracy/test_altaz_icrs.py]: Yeah, I'm also confused by this.  I just tried a merge of this branch locally and found that ``5`` still works just fine...?

Comment (taldcroft) [inline astropy/coordinates/tests/accuracy/test_altaz_icrs.py]: This was probably due to the iers-b issue.  I agree it should be possible to revert this change.

**Comment by eteq** [comment]
OK, I have some inkling of what's going on here: if you look at http://datacenter.iers.org/web/guest/eop you can see the original sources of the IERS_B files.  There you'll see the curious fact that the official IERS B bulletin and the one that `eopc04_IAU2000.62-now` comes from *also* don't agree (although well within the errors).  This is probably a smoothing effect because the whole thing comes from fitting models that depend on the data, so the small variations are probably due to wiggling polynomials and the like. I suspect the same holds for UT1-UTC.

The ``finals2000A`` file says it's updated weekly.  But the bulletin B is updated on different periods and the "real" underlying bulletin A is updated at yet a different time.  So my hunch is that sometimes they're not consistent because one has not been updated when the other has, and there's some wiggle room.

That's a long way of saying: the B bulletin is probably always more right.  It's also unclear where/when ``finals2000A`` gets their B values anyway, so what's done here is probably the best choice.

**Comment by eteq** [comment]
Oh, and one related note (relevant for the question of tying us to a download): @mhvk, is there a specific reason for getting IERS B from  http://hpiers.obspm.fr/iers/eop/eopc04/eopc04_IAU2000.62-now instead of http://datacenter.iers.org/eop/-/somos/5Rgv/latest/214 ?  The latter seem more "official" but hard to know...

File: astropy/utils/iers/iers.py [hunk 4]
```diff
@@ -443,5 +520,143 @@ def pm_source(self, i):
...
+            # Get the latest version
+            try:
+                clear_download_cache(conf.iers_auto_url)
+                filename = download_file(conf.iers_auto_url, cache=True)
+            except Exception as err:
+                # Issue a warning here, perhaps user is offline.  An exception
+                # will be raised downstream when actually trying to interpolate
+                # predictive values.
+                warn(AstropyWarning('failed to download {}: {}'
+                                    .format(conf.iers_auto_url, str(err))))
```
Comment (eteq) [inline astropy/utils/iers/iers.py]: I think this warning needs to be more verbose, along the lines of the old (removed in this PR) `_IERS_HINT`.  It should specify why the user might care (might compromise precision or lead to later errors) and also should tell the user how to try again if they re-connect their internet.  E.g.,
```
If you re-establish your internet and want to update to the latest IERS in this python session, you will need to do:
     >>> from astropy.utils import iers
     >>> iers.IERS_Auto.iers_table = iers.IERS_Auto.open()
```

File: astropy/coordinates/builtin_frames/utils.py [hunk 5]
```diff
@@ -85,7 +76,7 @@ def get_polar_motion(time):
 
     if np.any(status == iers.TIME_BEYOND_IERS_RANGE):
         wmsg = ('Tried to get polar motions for times after IERS data is '
-                'valid. Defaulting to polar motion from the 50-yr mean for those.' + _IERS_HINT)
+                'valid. Defaulting to polar motion from the 50-yr mean for those.')
```
Comment (eteq) [inline astropy/coordinates/builtin_frames/utils.py]: Maybe add to this warning something along the lines of "this may affect precision at the 10s of arcsec level"?  (Because that information was previously in the hint.)

**Comment by eteq** [comment]
@taldcroft - I had a moderately careful look and I like this. 

One top-level concern, though.  I notice that if I disconnect the internet and try this, I got a ~10 sec wait until I get a ``URLError: <urlopen error [Errno -3] Temporary failure in name resolution>``.   Do you see the same thing?  Not sure if there's any way to address it (It doesn't seem controllable with the timeout option in ``urlopen``), and not a deal-breaker anyway, but it's definitely worth trying to cut that down if there's any way to do so.

**Comment by pllim** [comment]
@eteq , theoretically, the timeout should be controlled by `astropy.utils.data.Conf.remote_timeout` value. However, its default value is 3, not 10.

**Comment by mhvk** [comment]
@eteq - on which version to use: it is very confusing! I remember being confused when I tried to find these files, and when I looked now, I again am confused. I think at the time I started at https://hpiers.obs[m.fr and this led me to "Reference C04 series; Each day since 1962" at
https://hpiers.obspm.fr/eop-pc/index.php?index=C04&lang=en

But it seems that since I implemented this a new web service has appeared, via http://www.iers.org/, which leads to the file you mention. Indeed, they also host IERS A, so perhaps both can be changed.
...
I just sent a note to IERS cc'd to you and @taldcroft. Hopefully, we will hear back on what is best practice...

**Comment by mhvk** [comment]
On the timeout, it seems specific to not being able to reach the DNS server. Is there a separate timeout for that? (i.e., separate from not being able to reach the data server given its IP address).

**Comment by mhvk** [comment]
On versions/implementation: given that IERS B values in IERS A seem to be slightly off, it seems we'll need to update IERS B too...

Also, I do wonder if we cannot rewrite it such that `IERS` itself does the combination of A and B (i.e., that we indeed avoid the move to `IERS_Auto` in `time` and `coordinates`). Right now, `IERS B` is chosen by setting `IERS.read = IERS_B.read`.  Need to think if a similar change or two might do the trick here.

p.s. While I like the overall structure as a `QTable` subclass, I'm not all that proud about I wrangled the two types together...

**Comment by pllim** [comment]
Re: timeout

I *think* what @eteq saw was `except urllib.error.URLError as e` block in `astropy.utils.data.download_file()`, which in this case, passes `remote_timeout` value to `with contextlib.closing(urllib.request.urlopen(remote_url, timeout=timeout)) as remote:` block. Does it make a different where in the line the connection is lost?

**Comment by eteq** [comment]
For completion: we just got an email from the IERS folks - it sounds like they actually said the original URLs are better because there's lag time for them to copy the files from the original URLs to the iers.org site... I'm guessing it just doesn't really matter very much so it's probably fine to leave as-is.  They said they'd "investigate" the differences, though.  I think for now what's done here is fine.

**Comment by eteq** [comment]
@mhvk - would you be OK with using the current version as-is, or doing what I suggested above of making `IERSAuto`->`IERS` and `IERS`->`IERSBase`?  I like your idea about combining them better, *but* that's not a project for "by the end of today", and it would be good if we could get this functionality in for 1.2...

**Comment by eteq** [comment]
Oh, and on similar lines, I think the 10-sec pause is still worth it for the auto-downloading feature.  So I'm ok with merging this now and we can consider how we might fix that in subsequent ("bug" fix) PRs.

**Comment by taldcroft** [comment]
@eteq - there are still some outstanding minor issues and comments to resolve, but I think the only biggish one that I see is the comment from @mhvk: "Overall, it looks good, though I want to think a bit more whether it isn't possible to make more minimal changes to IERS itself, keeping additions mostly within the IERS_Auto class."

So apart from that I think this is the final list:

- [ ] Test with iers.conf.max_auto_age = None.
- [ ] Time and coordinate docs should be updated to provide cross reference links.
- [ ] Change / remove the new IERS-A snippets.  I think it's possible to do the same testing without the 2 new snippet files I added.
- [ ] Squash things down.
- [ ] Add / improve 2 warnings as noted by @eteq.
- [ ] Revert 10 arcsec back to 5.

**Comment by taldcroft** [comment]
> Second, I'm a bit confused about the use of IERS_Auto everywhere where IERS was being used instead of just replacing IERS with IERS_Auto (IERS could be changed to IERSBase or something). Is there some compelling reason for that which I missed?

The reason was back-compatibility.  If there are users that are directly using `IERS` as currently documented, then making this change would break their code or at least change results.  I also think it is somewhat useful to have the new `Auto` word in there to make it clear that this class is a bit unusual in that it can auto-download a big file over the net.

**Comment by eteq** [comment]
Ah, gotcha - I was mostly thinking switching `IERSAuto`->`IERS` would just save some effort, but it's a good point that it's nominally an API break. So that all sounds good (also it needs a mostly-harmless rebase).  

On @mhvk's biggish point, I think if something can get done reasonably by today to address that then good, but it's better to get this in (with the changes you listed), rather than nothing at all for 1.2.

Also, FWIW, I just tried again on a non-interneted Mac (instead of Ubuntu in virtualbox like last time) and the download fails almost instantaneously.  So this is a platform-specific thing and I think probably not worth worrying about - the timeout as is should be fine.

**Comment by mhvk** [comment]
@eteq, @taldcroft - yes, I agree that this is a sufficiently huge improvement as is that we should merge something! Personally, I prefer the name change you suggested, although @taldcroft is right it is a somewhat of an API break. And there is something to be said for hinting at the "auto-download" aspect. So, @taldcroft, I think I'll just agree with whatever you decide.

As I mentioned, I do think I would like to reorganise a little, adding less auto-stuff to `IERS`, but think it can be done without breaking the public interface as defined now. So, as long as everybody is fine with possibly changing some of the private attributes, I think it is good to go.

In this respect: so far, the `iers` documentation has never been included, so I think it is OK to change some small things still, even if they are public. Might it be an idea to add a warning to the new iers doc that this is still somewhat in flux?

**Comment by taldcroft** [comment]
I think that the only functional change in `IERS` is the following, which just prevents linearly extrapolating outside the table range.  I *think* this is an improvement, but comments welcome.
```
           # Do not extrapolate outside range, instead just propagate last values.
           val[i == 0] = self[column][0]
           val[i == len(self)] = self[column][-1]
```
Otherwise the only changes in `IERS` are adding private methods that provide hooks within `_interpolate` that are needed by `IERS_Auto`.  The alternative would be essentially copying/reimplementing `_interpolate`.  Since these are all private I don't have any qualms about changing things later as needed.

Oh, if you want me to move `time_now()` into `IERS_A` and/or make it private that would be fine.  It probably should be private in any case.

**Comment by taldcroft** [comment]
One of the pains I have had during this process is the need to test with the `--remote-data` flag.  So one concern going forward is remembering to do this for development that uses `IERS_Auto`.  Without that flag then basically the tests are using `IERS_B` and so do not precisely replicate what the user will typically get.  I'm not sure if it should at least be a release process step as a final check.  Note that the coordinates `name_resolve` tests currently fail when run with `--remote-data` (#4834).

**Comment by eteq** [comment]
@taldcroft - hmm, I think the linear extrapolation is closer to what we *want*, though, right?  That is, it's better than just the last point?

**Comment by eteq** [comment]
Oh, wait, I now realize it's extrapolating *only* the last two.  Yeah, that's probably more noise than anything else.  It's kind of bike-shedding, though... Which makes me think it might be better to just leave in the old behavior, in case someone has written code expecting that?  But it probably doesn't matter much.

**Comment by eteq** [comment]
And FWIW, when I do releases I *always* run the tests both with and without ``--remote-data``.  Not sure if that's documented, but that's the procedure we tend to use (and it has indeed caught some last-minute bugs).

**Comment by pllim** [comment]
You might see something similar to #4896 with `--remote-data`.

**Comment by mhvk** [comment]
@taldcroft - I think not extrapolating is better (or setting to `0` -- it really doesn't matter and we do warn!). And yes, let's make `time_now` private.

**Comment by eteq** [comment]
I think setting to ``0`` is bad because it makes a sudden discontinuity in just-off-the-edge.  But I'm fine either way with extrapolating or not (i.e., +0 to leaving it like current master).

**Comment by taldcroft** [comment]
Extrapolating over long time scales would be bad (just imagine `PM_x` at JD=10000), and using the last value is closest for small extrapolations:
image

**Comment by eteq** [comment]
Good point re: long time scales. I'm convinced (just make sure to add a quick note about that in API changes, is this is a small but *possibly* important change for anyone who actually cares about it)

File: astropy/coordinates/tests/accuracy/test_altaz_icrs.py (hunk 3 above)
Comment (taldcroft) [inline astropy/coordinates/tests/accuracy/test_altaz_icrs.py]: Actually this is still a problem.  @eteq - maybe you ran tests without `--remote-data` (which falls back to the IERS-B values)?

I think the source of difference is that with `IERS_Auto` you get the last available values for `UT1-UTC` instead of zero (for out-of-range `IERS-B`).  That gives a time diff of 0.6 sec, which can mean up to 0.6 * 15 arcsec.

Looking at the testing for AltAz, I wonder if the discrepancies are related to time diffs?

Anyway, I propose that changing to the original 5 arcsec tolerance to 10 arcsec should be OK now that we understand it.

Comment (mhvk) [inline astropy/coordinates/tests/accuracy/test_altaz_icrs.py]: OK, that sounds good.  When you change the line, could you add a comment to remind us?

**Comment by taldcroft** [comment]
@eteq @mhvk - I think I have addressed all the comments in the last 6 commits.  I have to step away for dinner but will rebase and squash a bit after that.

**Comment by eteq** [comment]
@taldcroft - still needs a changelog entry (including my suggestion of an "API change" mention for the changed extrapolation), but maybe you meant to do that post-rebase.  If so, looks pretty solid!

**Comment by taldcroft** [comment]
@eteq - yes, I had forgotten about the changelog, but compensated now by putting in 4 separate entries!

Rebased version pushed.  The commit in 274a10a might be controversial.  I discovered that the `hor2eq` AltAz comparisons were failing again with `--remote-data`, and badly for the no-atmosphere case.  So I actually just ran hor2eq on our machine (fortunately have IDL and it just ran out of the box) for a more reasonable date, namely 2000-01-01, where definitive Earth rotation values are available.  These did not match at the ~5 arcsec level.  At this time I noticed that if I set the time scale of `obstime` to UT1 then the original tests (`< 5 arcsec` and `< 0.4 arcsec`) were passing for both `--remote-data` and not.

So I can't say *exactly* what is going on here (just too long working on this), except that I think that previously the tests were "accidentally" passing by virtue of effectively zeroing out `delta_ut1_utc` for the 2041 test date.

**Comment by eteq** [comment]
Huh, strange about the `hor2eq` business.  But I wouldn't worry too much - I noticed this in the "todo" section of `hor2eq`:

>  Difference between UTC and UT1 in determining LAST -- is this important?****

So my hunch is that `hor2eq` is implicitly assuming UTC=UT1, and that's why it worked in the `2041` case, just as you said.  So I think what you've done here is just fine.

If the tests pass, I say we merge!

**Comment by eteq** [comment]
🎆

**Comment by eteq** [comment]
backported to 1.2.x in 7048693cec1774179d267a5e4ff18668b20fd79d
